scripts/should_deploy.py

```diff
@@ -0,0 +1,247 @@
+        while href:
+            resp = self._request(href, params)
+    
+            for b in resp['builds']:
+                yield b
+    
+            try:
+                href = resp['@pagination']['next']['@href']
+            except KeyError:
+                raise StopIteration
+
+    def builds_for_branch(self, branch):
+        return self.builds(branch)
+
+    def builds_for_commit(self, branch, sha):
+        for b in self.builds(branch):
+            if b['commit']['sha'] == sha:
+                yield b
+
+
+class GitLab:
+    url = "https://allspark.dev.data.humancellatlas.org/api/v4"
+
+    def __init__(self, owner, repo):
+        self.owner = owner
+        self.repo = repo
+
+    def _request(self, href, params=None):
+        token = os.environ['GITLAB_TOKEN']
+        headers = {"PRIVATE-TOKEN": token}
+
+        if params is None:
+            params = dict()
+        
+        req_url = self.url + href
+        while req_url:
+            page = requests.get(req_url, headers=headers, params=params)
+            yield json.loads(page.text)
+            req_url = parse_paging_header(page.headers)
+
+    @max_items()
+    def builds(self, branch):
+        repo_slug = urllib.parse.quote_plus(f"{self.owner}/{self.repo}")
+        for page in self._request(f"/projects/{repo_slug}/jobs", {"ref_name": branch}):
+            for c in page:
+                yield c
+
+    @max_items()
+    def builds_for_commit(self, branch, sha):
+        for b in self.builds(branch):
+            if b['commit']['id'] == sha:
+                yield b
+
+    @max_items()
+    def commits(self, branch):
+        repo_slug = urllib.parse.quote_plus(f"{self.owner}/{self.repo}")
+        for page in self._request(f"/projects/{repo_slug}/repository/commits", {"ref_name": branch}):
+            for c in page:
+                yield c
+
+    def latest_commit(self, branch):
+        for c in self.commits(branch):
+            return c['id']
+
+
+def should_deploy_dev():
+    owner = "HumanCellAtlas"
+    repo = "data-store"
+    branch = "master"
+
+    GH = GitHub(owner, repo)
+    GL = GitLab(owner, repo)
+
+    gitlab_latest = GL.latest_commit(branch)
+    github_latest = GH.latest_commit(branch)
+
+    info = f"{owner} {repo} {branch}"
+
+    if gitlab_latest != github_latest:
+        raise Exception(f"GitHub and GitLab have different HEAD: {info} gitlab={gitlab_latest} github={github_latest}")
+
+    sha = gitlab_latest
+
+    for b in GL.builds_for_commit(branch, gitlab_latest):
+        raise Exception(f"Build already attempted: {info} {sha}")
+
+    integration_build_state = None
+    unit_tests_build_state = None
+    for b in Travis(owner, repo).builds_for_commit(branch, github_latest):
+        if b['commit']['message'].startswith("Integration"):
+            integration_build_state = b['state']
+        else:
+            unit_tests_build_state = b['state']
+
+        if integration_build_state is None and unit_tests_build_state is None:
```

Comment:
This looks odd. Don't you want to break out if both of them are *not* None?

Reply:
Insufficient coffee